File: yasmin_factory/test/test_yasmin_factory.cpp
// Copyright (C) 2025 Miguel Ángel González Santamarta
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

#include <fstream>
#include <gtest/gtest.h>
#include <memory>
#include <sstream>
#include <string>

#include "yasmin/blackboard/blackboard.hpp"
#include "yasmin/state.hpp"
#include "yasmin/state_machine.hpp"
#include "yasmin_factory/yasmin_factory.hpp"

using namespace yasmin_factory;

class TestYasminFactory : public ::testing::Test {
protected:
  std::unique_ptr<YasminFactory> factory;
  std::string test_dir;

  void SetUp() override {
    factory = std::make_unique<YasminFactory>();
    // Get the current test directory
    test_dir = __FILE__;
    test_dir = test_dir.substr(0, test_dir.find_last_of("/\\"));
  }

  void TearDown() override {
    if (factory) {
      factory->cleanup();
    }
  }

  // Helper function to create a temporary XML file
  std::string createTempXMLFile(const std::string &content) {
    std::string temp_file = test_dir + "/temp_test.xml";
    std::ofstream ofs(temp_file);
    ofs << content;
    ofs.close();
    return temp_file;
  }

  // Helper function to remove temporary file
  void removeTempFile(const std::string &filename) {
    std::remove(filename.c_str());
  }
};

TEST_F(TestYasminFactory, TestCreateCppState) {
  // Create a simple C++ state using XML
  std::string xml_content = R"(
    <State name="TestState" type="cpp" class="yasmin_factory/TestSimpleState"/>
  )";

  tinyxml2::XMLDocument doc;
  doc.Parse(xml_content.c_str());
  tinyxml2::XMLElement *state_elem = doc.FirstChildElement("State");

  ASSERT_NE(state_elem, nullptr);

  try {
    auto state = factory->create_state(state_elem);

    ASSERT_NE(state, nullptr);
    auto outcomes = state->get_outcomes();
    EXPECT_EQ(outcomes.size(), 2);
    EXPECT_NE(outcomes.find("outcome1"), outcomes.end());
    EXPECT_NE(outcomes.find("outcome2"), outcomes.end());
  } catch (const std::exception &e) {
    GTEST_SKIP() << "C++ plugin not available: " << e.what();
  }
}

TEST_F(TestYasminFactory, TestCreatePythonState) {
  // Create a simple Python state using XML
  std::string xml_content = R"(
    <State name="TestState" type="py" module="test.test_simple_state" class="TestSimpleState"/>
  )";

  tinyxml2::XMLDocument doc;
  doc.Parse(xml_content.c_str());
  tinyxml2::XMLElement *state_elem = doc.FirstChildElement("State");

  ASSERT_NE(state_elem, nullptr);

  try {
    auto state = factory->create_state(state_elem);

    ASSERT_NE(state, nullptr);
    auto outcomes = state->get_outcomes();
    EXPECT_EQ(outcomes.size(), 2);
  } catch (const std::exception &e) {
    GTEST_SKIP() << "Python state creation failed: " << e.what();
  }
}

TEST_F(TestYasminFactory, TestCreateStateInvalidType) {
  std::string xml_content = R"(
    <State name="InvalidState" type="invalid" class="SomeClass"/>
  )";

  tinyxml2::XMLDocument doc;
  doc.Parse(xml_content.c_str());
  tinyxml2::XMLElement *state_elem = doc.FirstChildElement("State");

  ASSERT_NE(state_elem, nullptr);

  EXPECT_THROW(
      { auto state = factory->create_state(state_elem); }, std::runtime_error);
}

TEST_F(TestYasminFactory, TestCreateStateFromFileCppPy) {
  std::string xml_file = test_dir + "/test_simple_sm_1.xml";

  try {
    auto sm = factory->create_sm_from_file(xml_file);

    ASSERT_NE(sm, nullptr);
    auto outcomes = sm->get_outcomes();
    EXPECT_NE(outcomes.find("end"), outcomes.end());
  } catch (const std::exception &e) {
    GTEST_SKIP() << "XML file not available or plugin missing: " << e.what();
  }
}

TEST_F(TestYasminFactory, TestCreateStateFromFilePyCpp) {
  std::string xml_file = test_dir + "/test_simple_sm_2.xml";

  try {
    auto sm = factory->create_sm_from_file(xml_file);

    ASSERT_NE(sm, nullptr);
    auto outcomes = sm->get_outcomes();
    EXPECT_NE(outcomes.find("end"), outcomes.end());
  } catch (const std::exception &e) {
    GTEST_SKIP() << "XML file not available or plugin missing: " << e.what();
  }
}

TEST_F(TestYasminFactory, TestCreateStateFromInvalidFile) {
  std::string xml_file = "/non/existent/file.xml";

  EXPECT_THROW(
      { auto sm = factory->create_sm_from_file(xml_file); },
      std::runtime_error);
}

TEST_F(TestYasminFactory, TestCreateStateFromInvalidXML) {
  std::string temp_file = createTempXMLFile("This is not valid XML");

  EXPECT_THROW(
      { auto sm = factory->create_sm_from_file(temp_file); },
      std::runtime_error);

  removeTempFile(temp_file);
}

TEST_F(TestYasminFactory, TestCreateStateFromInvalidRootElement) {
  std::string xml_content = R"(<?xml version="1.0"?>
    <InvalidRoot/>
  )";

  std::string temp_file = createTempXMLFile(xml_content);

  EXPECT_THROW(
      { auto sm = factory->create_sm_from_file(temp_file); },
      std::runtime_error);

  removeTempFile(temp_file);
}

TEST_F(TestYasminFactory, TestCreateSimpleStateMachine) {
  std::string xml_content = R"(
    <StateMachine outcomes="end">
      <State name="State1" type="cpp" class="yasmin_factory/TestSimpleState">
        <Transition from="outcome1" to="State1"/>
        <Transition from="outcome2" to="end"/>
      </State>
    </StateMachine>
  )";

  tinyxml2::XMLDocument doc;
  doc.Parse(xml_content.c_str());
  tinyxml2::XMLElement *root = doc.FirstChildElement("StateMachine");

  ASSERT_NE(root, nullptr);

  try {
    auto sm = factory->create_sm(root);

    ASSERT_NE(sm, nullptr);
    auto outcomes = sm->get_outcomes();
    EXPECT_NE(outcomes.find("end"), outcomes.end());
  } catch (const std::exception &e) {
    GTEST_SKIP() << "C++ plugin not available: " << e.what();
  }
}

TEST_F(TestYasminFactory, TestStateExecution) {
  std::string xml_content = R"(
    <State name="TestState" type="cpp" class="yasmin_factory/TestSimpleState"/>
  )";

  tinyxml2::XMLDocument doc;
  doc.Parse(xml_content.c_str());
  tinyxml2::XMLElement *state_elem = doc.FirstChildElement("State");

  ASSERT_NE(state_elem, nullptr);

  try {
    auto state = factory->create_state(state_elem);
    auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

    // Execute state multiple times
    std::string outcome1 = state->execute(blackboard);
    EXPECT_EQ(outcome1, "outcome1");
    EXPECT_EQ(blackboard->get<int>("counter"), 1);

    std::string outcome2 = state->execute(blackboard);
    EXPECT_EQ(outcome2, "outcome1");
    EXPECT_EQ(blackboard->get<int>("counter"), 2);

    std::string outcome3 = state->execute(blackboard);
    EXPECT_EQ(outcome3, "outcome2");
    EXPECT_EQ(blackboard->get<int>("counter"), 3);
  } catch (const std::exception &e) {
    GTEST_SKIP() << "C++ plugin not available: " << e.what();
  }
}

TEST_F(TestYasminFactory, TestNestedStateMachine) {
  std::string xml_file = test_dir + "/test_nested_sm.xml";

  try {
    auto sm = factory->create_sm_from_file(xml_file);

    ASSERT_NE(sm, nullptr);
    auto outcomes = sm->get_outcomes();
    EXPECT_NE(outcomes.find("final_outcome"), outcomes.end());
  } catch (const std::exception &e) {
    GTEST_SKIP() << "XML file not available or plugin missing: " << e.what();
  }
}

TEST_F(TestYasminFactory, TestFactoryCleanup) {
  std::string xml_content = R"(
    <State name="TestState" type="cpp" class="yasmin_factory/TestSimpleState"/>
  )";

  tinyxml2::XMLDocument doc;
  doc.Parse(xml_content.c_str());
  tinyxml2::XMLElement *state_elem = doc.FirstChildElement("State");

  try {
    auto state = factory->create_state(state_elem);
    ASSERT_NE(state, nullptr);

    // Cleanup should not raise errors
    factory->cleanup();

    // Should still be able to create new factory
    auto new_factory = std::make_unique<YasminFactory>();
    auto new_state = new_factory->create_state(state_elem);
    ASSERT_NE(new_state, nullptr);
    new_factory->cleanup();
  } catch (const std::exception &e) {
    GTEST_SKIP() << "C++ plugin not available: " << e.what();
  }
}

TEST_F(TestYasminFactory, TestMissingRequiredAttribute) {
  std::string xml_content = R"(
    <State name="NoClass" type="cpp"/>
  )";

  tinyxml2::XMLDocument doc;
  doc.Parse(xml_content.c_str());
  tinyxml2::XMLElement *state_elem = doc.FirstChildElement("State");

  ASSERT_NE(state_elem, nullptr);

  EXPECT_THROW(
      { auto state = factory->create_state(state_elem); }, std::runtime_error);
}

TEST_F(TestYasminFactory, TestRemapping) {
  std::string xml_file = test_dir + "/test_remapping.xml";

  try {
    auto sm = factory->create_sm_from_file(xml_file);
    ASSERT_NE(sm, nullptr);

    // Set up initial input
    auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();
    blackboard->set<std::string>("initial_data", "start");

    std::string outcome = (*sm)(blackboard);

    // Check the chain of processing
    EXPECT_EQ(outcome, "end");
    EXPECT_TRUE(blackboard->contains("final_data"));
    EXPECT_EQ(blackboard->get<std::string>("final_data"), "processed_start");
    EXPECT_TRUE(blackboard->contains("final_data_2"));
    EXPECT_EQ(blackboard->get<std::string>("final_data_2"),
              "processed_processed_start");
  } catch (const std::exception &e) {
    GTEST_SKIP() << "XML file not available or plugin missing: " << e.what();
  }
}

TEST_F(TestYasminFactory, TestFilePathMechanism) {
  std::string xml_file = test_dir + "/test_file_path_sm.xml";

  try {
    auto sm = factory->create_sm_from_file(xml_file);

    ASSERT_NE(sm, nullptr);

    // Verify the state machine name is set
    EXPECT_EQ(sm->get_name(), "MainStateMachine");

    // Verify the outcomes include the final outcome
    auto outcomes = sm->get_outcomes();
    EXPECT_NE(outcomes.find("final_end"), outcomes.end());

    // Execute the state machine
    auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();
    std::string outcome = (*sm)(blackboard);

    // The state machine should execute through FirstState -> IncludedSM ->
    // final_end or directly to final_end depending on FirstState's execution
    EXPECT_EQ(outcome, "final_end");
  } catch (const std::exception &e) {
    GTEST_SKIP() << "XML file not available or plugin missing: " << e.what();
  }
}

int main(int argc, char **argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: yasmin_factory/test/test_yasmin_factory.py
# Copyright (C) 2025 Miguel Ángel González Santamarta
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import unittest
import os
import tempfile
from lxml import etree as ET
from yasmin import Blackboard, StateMachine
from yasmin_factory.yasmin_factory import YasminFactory


class TestYasminFactory(unittest.TestCase):

    def setUp(self):
        """Set up test fixtures."""
        self.factory = YasminFactory()
        self.test_dir = os.path.dirname(os.path.abspath(__file__))

    def test_create_python_state_simple(self):
        """Test creating a simple Python state."""
        state_xml = '<State name="TestState" type="py" module="test.test_simple_state" class="TestSimpleState"/>'
        state_elem = ET.fromstring(state_xml)

        state = self.factory.create_state(state_elem)

        self.assertIsNotNone(state)
        self.assertEqual(len(state.get_outcomes()), 2)
        self.assertIn("outcome1", state.get_outcomes())
        self.assertIn("outcome2", state.get_outcomes())

    def test_create_cpp_state(self):
        """Test creating a C++ state using pybind bridge."""
        state_xml = (
            '<State name="CppState" type="cpp" class="yasmin_factory/TestSimpleState"/>'
        )
        state_elem = ET.fromstring(state_xml)

        state = self.factory.create_state(state_elem)
        self.assertIsNotNone(state)
        self.assertEqual(len(state.get_outcomes()), 2)

    def test_create_state_invalid_type(self):
        """Test creating a state with invalid type."""
        state_xml = '<State name="InvalidState" type="invalid" class="SomeClass"/>'
        state_elem = ET.fromstring(state_xml)

        with self.assertRaises(ValueError) as context:
            self.factory.create_state(state_elem)

        self.assertIn("Unknown state type", str(context.exception))

    def test_create_state_missing_class(self):
        """Test creating a state with missing class attribute."""
        state_xml = '<State name="NoClass" type="py" module="some.module"/>'
        state_elem = ET.fromstring(state_xml)

        with self.assertRaises(KeyError):
            self.factory.create_state(state_elem)

    def test_create_state_missing_module(self):
        """Test creating a Python state with missing module attribute."""
        state_xml = '<State name="NoModule" type="py" class="SomeClass"/>'
        state_elem = ET.fromstring(state_xml)

        with self.assertRaises(KeyError):
            self.factory.create_state(state_elem)

    def test_create_sm_from_xml(self):
        """Test creating a state machine from XML string."""
        sm_xml = """
        <StateMachine outcomes="end">
            <State name="State1" type="py" module="test.test_simple_state" class="TestSimpleState">
                <Transition from="outcome1" to="State1"/>
                <Transition from="outcome2" to="end"/>
            </State>
        </StateMachine>
        """
        root = ET.fromstring(sm_xml)

        sm = self.factory.create_sm(root)

        self.assertIsNotNone(sm)
        self.assertIsInstance(sm, StateMachine)
        self.assertIn("end", sm.get_outcomes())

    def test_create_sm_from_file_cpppy(self):
        """Test creating a state machine from XML file."""
        xml_file = os.path.join(self.test_dir, "test_simple_sm_1.xml")
        sm = self.factory.create_sm_from_file(xml_file)

        self.assertIsNotNone(sm)
        self.assertIsInstance(sm, StateMachine)
        self.assertIn("end", sm.get_outcomes())

    def test_create_sm_from_file_pycpp(self):
        """Test creating a state machine from XML file."""
        xml_file = os.path.join(self.test_dir, "test_simple_sm_2.xml")
        sm = self.factory.create_sm_from_file(xml_file)

        self.assertIsNotNone(sm)
        self.assertIsInstance(sm, StateMachine)
        self.assertIn("end", sm.get_outcomes())

    def test_create_sm_from_invalid_file(self):
        """Test creating a state machine from non-existent file."""
        with self.assertRaises(Exception):
            self.factory.create_sm_from_file("/non/existent/file.xml")

    def test_create_sm_from_invalid_xml(self):
        """Test creating a state machine from invalid XML."""
        with tempfile.NamedTemporaryFile(mode="w", suffix=".xml", delete=False) as f:
            f.write("This is not valid XML")
            temp_file = f.name

        try:
            with self.assertRaises(Exception):
                self.factory.create_sm_from_file(temp_file)
        finally:
            os.unlink(temp_file)

    def test_create_sm_invalid_root_element(self):
        """Test creating a state machine with invalid root element."""
        with tempfile.NamedTemporaryFile(mode="w", suffix=".xml", delete=False) as f:
            f.write('<?xml version="1.0"?><InvalidRoot/>')
            temp_file = f.name

        try:
            with self.assertRaises(ValueError) as context:
                self.factory.create_sm_from_file(temp_file)
            self.assertIn("Root element must be 'StateMachine'", str(context.exception))
        finally:
            os.unlink(temp_file)

    def test_state_execution(self):
        """Test executing a created state."""
        state_xml = '<State name="TestState" type="py" module="test.test_simple_state" class="TestSimpleState"/>'
        state_elem = ET.fromstring(state_xml)

        state = self.factory.create_state(state_elem)
        blackboard = Blackboard()

        # Execute state multiple times
        outcome1 = state.execute(blackboard)
        self.assertEqual(outcome1, "outcome1")
        self.assertEqual(blackboard["counter"], 1)

        outcome2 = state.execute(blackboard)
        self.assertEqual(outcome2, "outcome1")
        self.assertEqual(blackboard["counter"], 2)

        outcome3 = state.execute(blackboard)
        self.assertEqual(outcome3, "outcome2")
        self.assertEqual(blackboard["counter"], 3)

    def test_nested_state_machine(self):
        """Test creating a nested state machine."""
        xml_file = os.path.join(self.test_dir, "test_nested_sm.xml")
        sm = self.factory.create_sm_from_file(xml_file)

        self.assertIsNotNone(sm)
        self.assertIsInstance(sm, StateMachine)
        self.assertIn("final_outcome", sm.get_outcomes())

    def test_remapping(self):
        """Test remapping across multiple states that pass data through the chain."""
        xml_file = os.path.join(self.test_dir, "test_remapping.xml")
        sm = self.factory.create_sm_from_file(xml_file)

        self.assertIsNotNone(sm)

        # Set up initial input
        blackboard = Blackboard()
        blackboard["initial_data"] = "start"

        outcome = sm(blackboard)

        # Check the chain of processing
        self.assertEqual(outcome, "end")
        self.assertTrue(blackboard.contains("final_data"))
        self.assertEqual(blackboard.get("final_data"), "processed_start")
        self.assertTrue(blackboard.contains("final_data_2"))
        self.assertEqual(blackboard.get("final_data_2"), "processed_processed_start")

    def test_file_path_mechanism(self):
        """Test including external state machine via file_path attribute."""
        xml_file = os.path.join(self.test_dir, "test_file_path_sm.xml")
        sm = self.factory.create_sm_from_file(xml_file)

        self.assertIsNotNone(sm)
        self.assertIsInstance(sm, StateMachine)

        # Verify the state machine name is set
        self.assertEqual(sm.get_name(), "MainStateMachine")

        # Verify the outcomes include the final outcome
        self.assertIn("final_end", sm.get_outcomes())

        # Execute the state machine
        blackboard = Blackboard()
        outcome = sm(blackboard)

        # The state machine should execute through FirstState -> IncludedSM -> final_end
        # or directly to final_end depending on FirstState's execution
        self.assertIn(outcome, ["final_end"])


if __name__ == "__main__":
    unittest.main()
